File: LeetCode/70. Climbing Stairs.py
#https://leetcode.com/problems/climbing-stairs/
class Solution:
    def climbStairs(self, n: 'int') -> 'int':
        if n <= 2:
            return n

        first = 1
        second = 2

        for i in range(3, n):
            temp = first + second
            first, second = second, temp

        return first + second

File: LeetCode/70. Climbing Stairs.java
// https://leetcode.com/problems/climbing-stairs/

import java.util.HashMap;

class Solution {
    private HashMap<Integer, Integer> memo = new HashMap<>();
    
    public int climbStairs(int n) {
        //pattern is fibonacci
        return calculate(n);
    }
    
    private int calculate(int n) {
        if(n <= 2) {
            return n;
        }
        
        if(memo.containsKey(n)) return memo.get(n);
        
        int result = 0;
        result = calculate(n-1) + calculate(n-2);
        memo.put(n, result);
        return result;
    }
}
